File: app/src/main/cpp/wn-steam-client/src/pb/cmsg_client_playing_session_state.cpp
#include "wn_steam/pb/cmsg_client_playing_session_state.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

std::optional<CMsgClientPlayingSessionState>
CMsgClientPlayingSessionState::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CMsgClientPlayingSessionState m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 2:
                if (auto v = r.boolean(); v) m.playing_blocked = *v;
                else return std::nullopt;
                break;
            case 3:
                if (auto v = r.u32(); v) m.playing_app = *v;
                else return std::nullopt;
                break;
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cmsg_client_playing_session_state.rs
use crate::proto_wire::Reader;

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CMsgClientPlayingSessionState {
    pub playing_blocked: bool,
    pub playing_app: u32,
}

impl CMsgClientPlayingSessionState {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                2 => msg.playing_blocked = reader.boolean()?,
                3 => msg.playing_app = reader.u32()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::pb::cmsg_client_change_status::CMsgClientChangeStatus;
    use crate::pb::cmsg_client_kick_playing_session::CMsgClientKickPlayingSession;
    use crate::proto_wire::Writer;

    #[test]
    fn force_emits_zero_and_false_values() {
        assert_eq!(
            CMsgClientChangeStatus::default().serialize(),
            vec![0x08, 0x00]
        );
        assert_eq!(
            CMsgClientKickPlayingSession {
                only_stop_game: false
            }
            .serialize(),
            vec![0x08, 0x00]
        );
    }

    #[test]
    fn parses_playing_session_state() {
        let mut body = Vec::new();
        let mut w = Writer::new(&mut body);
        w.bool_field(2, true);
        w.uint32_field(3, 480);
        let parsed = CMsgClientPlayingSessionState::deserialize(&body).unwrap();
        assert!(parsed.playing_blocked);
        assert_eq!(parsed.playing_app, 480);
    }
}
